### PR TITLE
Preparing run2 legacy v4 part2

### PR DESCRIPTION
**Pull request by jedori0228**
- Muon MVA variables added : https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideMuonIdRun2#Lepton_MVA
- Muon miniiso variables added; muon_jetPtRatio and muon_jetPtRel. No function available for electrons though.. We should calculate them ourselves later, but saving this will be helpful for the cross check between POG and us.
- Sim-matching variables added : https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideMuonIdRun2#SimHit_Muon_matching_since_CMSSW;

### CODE REVIEW

**Comment by jedori0228** [comment]
LHE object added

**Comment by jedori0228** [comment]
EGamma objects has userInt('VID NAME') which has all the pass/fail informations of each cuts. This is very useful when using 'Loose w/o iso'.

**Comment by jedori0228** [comment]
This is the end of my Part2 update, so please do the review

**Comment by jedori0228** [comment]
I have one more PR.. Please review this ASAP

File: SKFlatMaker/script/varlist_LHE.txt [hunk 1]
```diff
@@ -1,3 +1,9 @@
 double	PDFWeights_Scale	1
 double	PDFWeights_Error	1
 double	PDFWeights_AlphaS	1
+double	LHE_Px	1
+double	LHE_Py	1
+double	LHE_Pz	1
+double	LHE_E	1
+double	LHE_Status	1
+double	LHE_ID	1
```
Comment (bhoh) [inline SKFlatMaker/script/varlist_LHE.txt]: I think **int** type is enough for **LHE_Status**, **LHE_ID** variables.

File: SKFlatMaker/interface/SKFlatMaker.h [hunk 2]
```diff
@@ -634,12 +635,30 @@ class SKFlatMaker : public edm::EDAnalyzer
...
   RoccoR rc;
 
   //==== LHE
 
+  vector<double> LHE_Px;
+  vector<double> LHE_Py;
+  vector<double> LHE_Pz;
+  vector<double> LHE_E;
+  vector<double> LHE_Status;
+  vector<double> LHE_ID;
```
Comment (bhoh) [inline SKFlatMaker/interface/SKFlatMaker.h]: I think **int** type is enough for **LHE_Status**, **LHE_ID** variables.

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 3]
```diff
@@ -1060,11 +1080,30 @@ void SKFlatMaker::beginJob()
...
   
   // -- LHE info -- //
   if( theStoreLHEFlag ){
+
+    DYTree->Branch("LHE_Px", "vector<double>", &LHE_Px);
+    DYTree->Branch("LHE_Py", "vector<double>", &LHE_Py);
+    DYTree->Branch("LHE_Pz", "vector<double>", &LHE_Pz);
+    DYTree->Branch("LHE_E", "vector<double>", &LHE_E);
+    DYTree->Branch("LHE_Status", "vector<double>", &LHE_Status);
+    DYTree->Branch("LHE_ID", "vector<double>", &LHE_ID);
```
Comment (bhoh) [inline SKFlatMaker/src/SKFlatMaker.cc]: I think **int** type is enough for **LHE_Status**, **LHE_ID** variables.

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 4]
```diff
@@ -1816,6 +1855,25 @@ void SKFlatMaker::fillMuons(const edm::Event &iEvent, const edm::EventSetup& iSe
...
+
+    //==== Muon sim-matching variable
+    //==== https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideMuonIdRun2#SimHit_Muon_matching_since_CMSSW
+    muon_simType.push_back( imuon.simType() );
+    muon_simExtType.push_back( imuon.simExtType() );
+    muon_simFlavour.push_back( imuon.simFlavour() );
+    muon_simHeaviestMotherFlavour.push_back( imuon.simHeaviestMotherFlavour() );
+    muon_simPdgId.push_back( imuon.simPdgId() );
+    muon_simMotherPdgId.push_back( imuon.simMotherPdgId() );
+    //muon_simMatchQuality.push_back( imuon.simMatchQuality() );  // TODO not supported in CMSSW_10_2_10
```
Comment (bhoh) [inline SKFlatMaker/src/SKFlatMaker.cc]: As I checked twiki instruction, it seems fine.

File: SKFlatMaker/src/SKFlatMaker.cc [hunk 5]
```diff
@@ -2245,6 +2303,14 @@ el->deltaEtaSuperClusterTrackAtVtx() - el->superCluster()->eta() + el->superClus
...
         IDBit &= ~(1 << it_ID);
       }
+
+      //==== bits for each cuts
+      //==== https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideCMSPhysicsObjectSchoolAACHEN2019EGamma, Exercise 2
+      int this_idcutbid = 0;
+      if( el->hasUserInt(electron_IDtoSave.at(it_ID)) ){
+        this_idcutbid = el->userInt(electron_IDtoSave.at(it_ID));
+      }
+      electron_IDCutBit.push_back( this_idcutbid );
```
Comment (bhoh) [inline SKFlatMaker/src/SKFlatMaker.cc]: If **hasUserInt** returns **false**, 0 is inserted to **electron_IDCutBit**. But, I think it possibly cause a mistake when we run analyzer. how about insert a value like -1 when **hasUserInt** returns **false**.

**Comment by jedori0228** [comment]
Thanks @bhoh , I updated them!

**Pull request merged**